jung-api/src/main/java/edu/uci/ics/jung/graph/util/TreeUtils.java

```diff
@@ -105,11 +105,14 @@
    *
    * @param <N> the node type
    * @param <E> the edge type
-   * @param tree the tree to which {@code subTree} is to be added
+   * @param tree the tree to which {@code subTree} is to be added; must contain no nodes if {@code
+   *     subTreeParent} is {@code null}
    * @param subTree the tree which is to be grafted on to {@code tree}
    * @param subTreeParent the parent of the root of {@code subTree} in its new position in {@code
-   *     tree}
-   * @param connectingEdge the edge used to connect {@code subTreeParent} to {@code subtree}'s root
+   *     tree}; can be {@code null} to represent "no parent" i.e. if {@code tree} contains no node;
+   *     must be {@code null} if {@code tree} contains no nodes
+   * @param connectingEdge the edge used to connect {@code subTreeParent} to {@code subtree}'s root;
+   *     must be {@code null} if {@code subTreeParent} is {@code null}
    */
   public static <N, E> void addSubTree(
```

Comment:
The problem that we're trying to fix here is that we're trying to support the case where the "subtree" that's being added is in fact the entire tree (the entirety?  :) ).

I'm not sure that's actually a good idea, because it makes the behavior overly complex; it requires adding special-case behavior that connects three of the parameters.

We could create a simpler overload that would do what we want, but it would basically look like this:

    addSubTree(MutableCTreeNetwork tree, CTreeNetwork subTree) { ... }

and would still require that`tree` be empty.

What do you think of just documenting that the caller needs to do the null check on "parent"?

Side point: AFAICT the current version of`TreeCollapseDemo` doesn't call `addSubTree()`, or anything else from TreeUtils; where is that breakage actually occurring, or is it already otherwise resolved?

Reply:
Oops, sorry for not responding sooner @jrtom!

Ah yes, I see where you're coming from here. I agree with hindsight that having this special case behaviour makes things overly complex, so I'll go back and change things so that users cannot pass a null `subTreeParent` (and by extension cannot pass an empty `tree` either).

Regarding `TreeCollapseDemo`, the call to `addSubTree()` is actually cleverly hidden: `TreeCollapseDemo::new` calls `TreeCollapser::expand`, which in turn calls `TreeUtils::addSubTree`. So the breakage is actually in `TreeCollapser`, not `TreeCollapseDemo` directly. :)